sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/Refcounted.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.pubsublite.internal;
+
+import com.google.cloud.pubsublite.internal.wire.SystemExecutors;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** A class that safely ensures an object of type T is cleaned up before it is garbage collected. */
```

Comment:
I may be missing something, but I'm not so sure this name and description are accurate. I can see how this is a box that closes the thing it is holding, but that seems pretty different to the description IMO.

Reply:
The description is accurate. finalize() on Object runs when the object is garbage collected before its memory is freed. This wrapper ensures that the underlying object has close() called on it without indefinitely extending its lifetime beyond that of the refcounted wrapper.

Reply:
I don't believe "refcounted" is an accurate description for this, what with there not being any counting going on.

Reply:
I hate to delay things on naming, but this time it matters enough, I think. Perhaps `CloserReference` or some such. It is very similar to `Cleaner` but for `AutoCloseable`.

Reply:
Done.